Touch handler: Fix race condition and rate-limit move events

kongbai1996 opened the pull request:
* TouchSequenceInfo is added to store information about a touch sequence. In order to respect `preventDefault` we may need to wait until the touch handler has finished processing, which may be after a touch sequence has ended, and a new sequence has begun. In the previous code we had just a single state in the touch handler, which would cause issues when receiving back processed events from previous sequences. To avoid this race, we make the state be per-touch sequence
 
* The handling_touch_move attribute is added to the TouchHandler, indicating that the script is processing the touch move event. We use this information to rate-limit the amount of move-events we send to script, meaning script receives one move event at a time. The spec leaves the frequency of sending move events up to the implementation. 
 
* The first touch_move event is merged with all following tuch_move events, until we know if moving is allowed or not. This implementation should be improved in future work.


---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix a part of #35436  (GitHub issue number if applicable)

kongbai1996 committed:
TouchSequenceInfo is added to store information about a touch sequence.
  For details about TouchSequenceInfo, see the code comments.
The handling_touch_move attribute is added to the TouchHandler, indicating that the script is processing the touch move event.
  When handling_touch_move is set to true, the touch move event does not need to be sent to the script thread.

Signed-off-by: kongbai1996 <1782765876@qq.com>

kongbai1996 commented:
@jschwe please review, thanks.

kongbai1996 committed:
move touch state, active_touch_point and handling_touch_move to TouchSequenceInfo form TouchHandler.
remove TouchSequenceInfo end_sequence property, add Finished state mark sequence end.
if preventDefault on touchup, do not prevent Fling.

Signed-off-by: kongbai1996 <1782765876@qq.com>

jschwe commented on components/compositing/touch.rs:
```diff
@@ -22,10 +24,106 @@ const FLING_MIN_SCREEN_PX: f32 = 3.0;
+    // Cancel click when touch move occurs or multiple touches on current thread.
+    cancel_click: bool,
+    // prevent default action in the touch sequence
+    prevent_default: bool,
```
If everything is prevented, I think we can use TouchSequenceState::PreventDefault, so we don't waste any time doing useless calculations that will never be used anyway.

kongbai1996 replied:
If the status is set to preventdefault, subsequent panning operations will not be recognized. This is an error. preventdefault only prevents default operations. The move event needs to be sent to the script.

jschwe replied:
Additional context after offline discussion: We only send the move event to script if it covers more than the minimum distance, hence we still want to process move actions.

jschwe commented on components/compositing/touch.rs:
```diff
@@ -22,10 +24,106 @@ const FLING_MIN_SCREEN_PX: f32 = 3.0;
+    prevent_default: bool,
+    // Once the first move has been processed by script, we can transition to
+    // non-cancellable events, and directly perform the pan without waiting for script.
+    first_move_processed: bool,
```
Maybe we can remove this bool variable and instead have
`TouchSequenceState::WaitingForMoveHandler` and `TouchSequenceState::Moving` ?

kongbai1996 replied:
It's not easy to change.
If the Up state changes to Finish and then the MovePreocessed message is received, it is not possible to know whether the first Move is processed.

jschwe commented on components/compositing/touch.rs:
```diff
@@ -22,10 +24,106 @@ const FLING_MIN_SCREEN_PX: f32 = 3.0;
+    // non-cancellable events, and directly perform the pan without waiting for script.
+    first_move_processed: bool,
+    // Move operation waiting to be processed in the touch sequence
+    pending_touch_move_action: Option<TouchAction>,
```
If we have `TouchSequenceState::WaitingForMoveHandler`, we could move this state into the enum variant.

jschwe replied:
Let's leave the touch move for later, we need to think more about how to handle merging pending move actions.

jschwe commented on components/compositing/touch.rs:
```diff
@@ -22,10 +24,106 @@ const FLING_MIN_SCREEN_PX: f32 = 3.0;
+    // Move operation waiting to be processed in the touch sequence
+    pending_touch_move_action: Option<TouchAction>,
+    // Up operation waiting to be processed in the touch sequence
+    pending_touch_up_action: Option<TouchAction>,
```
Assuming `move` was allowed - Does `preventDefault()` in the `touch_up` handler prevent Flinging in other browsers? Or is it just a potential click that would be prevented.

kongbai1996 replied:
Fling won't be stopped. I'll fix it.

kongbai1996 commented on components/compositing/compositor.rs:
```diff
@@ -1478,69 +1483,140 @@ impl IOCompositor {
+            EventResult::DefaultPrevented(sequence_id, event_type) => {
+                self.touch_handler.prevent_default(sequence_id);
+                // if preventdefault on any where, remove pending_touch_up_action,
+                // It means that `TouchAction::Click` and `TouchAction::Flinging` won't happen.
```
Touchstart and touchmove prevent click, fling, scroll, and pinch operations.
touchup prevented only prevents click.

kongbai1996 commented on components/compositing/touch.rs:
```diff
@@ -22,10 +24,106 @@ const FLING_MIN_SCREEN_PX: f32 = 3.0;
+    // touch sequence active touch points
+    active_touch_points: Vec<TouchPoint>,
+    // Mark the end of the touch sequence
+    sequence_end: bool,
```
Modified

jschwe committed:
Refactor Touchhandler

- Add a newtype wrapper for the TouchSequenceId
- Move more state back into the TouchSequenceState
- Rename TouchAction to TouchMoveAction,
  since it only covers immediate actions now.
  Everything else is handled via state, since
  it needs to wait on the handler.

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

jschwe committed:
Fix test-tidy

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

jschwe committed:
Fix clippy missing-default lint

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

jschwe committed:
Fix remaining clippy lints

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

jschwe committed:
Remove accidental committed test file

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

jschwe committed:
Remove wrong todo comment

(move events that are sent to script are just raw  touchpoints,
 no merging needed)

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

jschwe reviewed (approved):
Thanks! I edited the PR with some suggestions (after offline discussions) and updated the description

jschwe committed:
Fix preventdefault after long touch_down handler

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>
